docs/spark/custom-key.md

```diff
@@ -0,0 +1,112 @@
+Writing a Custom Key Type
+=========================
+
+*Want to jump straight to a code example? See:*
+```
+doc-examples/src/main/scala/geotrellis/doc/examples/spark/VoxelKey.scala
+```
+
+Intro
+-----
+
+Keys are used to index (or "give a position to") tiles in a tile layer.
+Typically these tiles are arranged in some conceptual grid, for instance in
+a two-dimensional matrix via a `SpatialKey`. There is also a `SpaceTimeKey`,
+which arranges tiles in a cube of two spatial dimensions and one time
+dimension.
+
+In this way, keys define how a tile layer is shaped. Here, we provide an example of how
+to define a new key type, should you want a custom one for your application.
+
+The `VoxelKey` type
+-------------------
+
+A voxel is the 3D analogue to a 2D pixel. By defining a new `VoxelKey` type,
+we can create grids of tiles that have a 3D spatial relationship. The class
+definition itself is simple:
+
+```scala
+case class VoxelKey(x: Int, y: Int, z: Int)
+```
+
+Key usage in many GeoTrellis operations is done generically with a `K` type
+parameter, for instance in the `S3LayerReader` class:
+
+```scala
+/* Read a tile layer from S3 via a given `LayerId`. Function signature slightly simplified. */
+S3LayerReader.read[K: Boundable: JsonFormat, V, M]: LayerId => RDD[(K, V)] with Metadata[M]
+```
+
+Where the pattern `[A: Trait1: Trait2: ...]` means that for whichever `A`
+you end up using, it must have an implicit instance of `Trait1` and `Trait2`
```

Comment:
`Trait1[K]` and `Trait2[K]` as `[A: Trait1: Trait2: ...]` expands to `[A](ev0: Trait1[A], ev1: Trait2[A])`